api/sms_provider/sms_provider_test.go

```diff
@@ -0,0 +1,216 @@
+					AuthToken:         "test_auth_token",
+					MessageServiceSid: "test_message_service_id",
+				},
+				Messagebird: conf.MessagebirdProviderConfiguration{
+					AccessKey:  "test_access_key",
+					Originator: "test_originator",
+				},
+				Vonage: conf.VonageProviderConfiguration{
+					ApiKey:    "test_api_key",
+					ApiSecret: "test_api_secret",
+					From:      "test_from",
+				},
+				Textlocal: conf.TextlocalProviderConfiguration{
+					ApiKey: "test_api_key",
+					Sender: "test_sender",
+				},
+			},
+		},
+	}
+	suite.Run(t, ts)
+}
+
+func (ts *SmsProviderTestSuite) TestTwilioSendSms() {
+	defer gock.Off()
+	provider, err := NewTwilioProvider(ts.Config.Sms.Twilio)
+	require.NoError(ts.T(), err)
+
+	twilioProvider, ok := provider.(*TwilioProvider)
+	require.Equal(ts.T(), true, ok)
+
+	phone := "123456789"
+	message := "This is the sms code: 123456"
+
+	body := url.Values{
+		"To":      {"+" + phone},
+		"Channel": {"sms"},
+		"From":    {twilioProvider.Config.MessageServiceSid},
+		"Body":    {message},
+	}
+
+	cases := []struct {
+		Desc           string
+		TwilioResponse *gock.Response
+		ExpectedError  error
+	}{
+		{
+			Desc: "Successfully sent sms",
+			TwilioResponse: gock.New(twilioProvider.APIPath).Post("").
+				MatchHeader("Authorization", "Basic "+base64.StdEncoding.EncodeToString([]byte(twilioProvider.Config.AccountSid+":"+twilioProvider.Config.AuthToken))).
+				MatchType("url").BodyString(body.Encode()).
+				Reply(200).JSON(SmsStatus{
+				To:     "+" + phone,
+				From:   twilioProvider.Config.MessageServiceSid,
+				Status: "sent",
+				Body:   message,
+			}),
+			ExpectedError: nil,
+		},
+		{
+			Desc: "Sms status is failed / undelivered",
+			TwilioResponse: gock.New(twilioProvider.APIPath).Post("").
+				MatchHeader("Authorization", "Basic "+base64.StdEncoding.EncodeToString([]byte(twilioProvider.Config.AccountSid+":"+twilioProvider.Config.AuthToken))).
+				MatchType("url").BodyString(body.Encode()).
+				Reply(200).JSON(SmsStatus{
+				ErrorMessage: "failed to send sms",
+				ErrorCode:    "401",
+				Status:       "failed",
+			}),
+			ExpectedError: fmt.Errorf("twilio error: %v %v", "failed to send sms", "401"),
+		},
+		{
+			Desc: "Non-2xx status code returned",
+			TwilioResponse: gock.New(twilioProvider.APIPath).Post("").
+				MatchHeader("Authorization", "Basic "+base64.StdEncoding.EncodeToString([]byte(twilioProvider.Config.AccountSid+":"+twilioProvider.Config.AuthToken))).
+				MatchType("url").BodyString(body.Encode()).
+				Reply(500).JSON(twilioErrResponse{
+				Code:     500,
+				Message:  "Internal server error",
+				MoreInfo: "error",
+				Status:   500,
+			}),
+			ExpectedError: &twilioErrResponse{
+				Code:     500,
+				Message:  "Internal server error",
+				MoreInfo: "error",
+				Status:   500,
+			},
+		},
+	}
+
+	for _, c := range cases {
+		ts.Run(c.Desc, func() {
+			err = twilioProvider.SendSms(phone, message)
+			require.Equal(ts.T(), c.ExpectedError, err)
+		})
+	}
```

Comment:
Another good check at the end of a test is `assert.Empty(t, gock.GetUnmatchedRequests())`. It ensures that all expectations you've set with gock are called once.

I also find its default logging a little unclear, so I created a [helper](https://github.com/supabase/cli/blob/main/internal/testing/apitest/helper.go) to print out unmatched requests.